// web/src/components/ExportButtons.jsx
import React, { useState } from 'react';
import { format } from 'date-fns';
import { Button } from '@/components/ui/button';
import { Download, FileText, Loader2 } from 'lucide-react';
import jsPDF from 'jspdf';
import 'jspdf-autotable';
import { toast } from 'sonner';

const ExportButtons = ({ transactions, metrics, userName }) => {
  const [isExportingCSV, setIsExportingCSV] = useState(false);
  const [isExportingPDF, setIsExportingPDF] = useState(false);

  const handleExportCSV = async () => {
    if (!transactions || transactions.length === 0) {
      toast.error('No data to export');
      return;
    }
    
    setIsExportingCSV(true);
    try {
      const headers = ['Date', 'Category', 'Description', 'Type', 'Amount', 'Running Balance', 'Source'];
      const rows = transactions.map(tx => [
        format(new Date(tx.date), 'yyyy-MM-dd'),
        tx.category || 'Manual',
        `"${(tx.description || '').replace(/"/g, '""')}"`,
        tx.transaction_type === 'debit' ? 'Debit' : 'Credit',
        tx.amount,
        tx.running_balance,
        tx.source_module || 'Manual'
      ]);
      
      const summaryRows = [
        [],
        ['SUMMARY'],
        ['Opening Balance', '', '', '', metrics.opening_balance],
        ['Total Income', '', '', '', metrics.total_income],
        ['Total Outflow', '', '', '', metrics.total_outflow],
        ['Closing Balance', '', '', '', metrics.net_balance]
      ];

      const csvContent = [
        headers.join(','), 
        ...rows.map(r => r.join(',')),
        ...summaryRows.map(r => r.join(','))
      ].join('\n');
      
      const blob = new Blob([csvContent], { type: 'text/csv;charset=utf-8;' });
      const link = document.createElement('a');
      link.href = URL.createObjectURL(blob);
      link.download = `cashbook_export_${format(new Date(), 'yyyyMMdd')}.csv`;
      link.click();
      toast.success('CSV exported successfully');
    } catch (error) {
      toast.error('Failed to export CSV');
    } finally {
      setIsExportingCSV(false);
    }
  };

  const handleExportPDF = async () => {
    if (!transactions || transactions.length === 0) {
      toast.error('No data to export');
      return;
    }

    setIsExportingPDF(true);
    try {
      const doc = new jsPDF();
      
      // Header
      doc.setFontSize(20);
      doc.setTextColor(15, 23, 42);
      doc.text('Cashbook Statement', 14, 22);
      
      doc.setFontSize(10);
      doc.setTextColor(100, 116, 139);
      doc.text(`Generated by: ${userName || 'Admin'}`, 14, 30);
      doc.text(`Date: ${format(new Date(), 'dd MMM yyyy HH:mm')}`, 14, 35);
      
      // Summary Box
      doc.setFillColor(248, 250, 252);
      doc.rect(14, 40, 182, 25, 'F');
      doc.setTextColor(15, 23, 42);
      doc.setFontSize(10);
      doc.text(`Opening Balance: Rs. ${metrics.opening_balance.toFixed(2)}`, 20, 48);
      doc.text(`Total Income: Rs. ${metrics.total_income.toFixed(2)}`, 20, 58);
      doc.text(`Total Outflow: Rs. ${metrics.total_outflow.toFixed(2)}`, 110, 48);
      doc.setFont(undefined, 'bold');
      doc.text(`Closing Balance: Rs. ${metrics.net_balance.toFixed(2)}`, 110, 58);
      doc.setFont(undefined, 'normal');

      // Table
      const tableColumn = ["Date", "Category", "Description", "Type", "Amount", "Balance"];
      const tableRows = transactions.map(tx => [
        format(new Date(tx.date), 'dd/MM/yyyy'),
        tx.category || 'Manual',
        tx.description || '-',
        tx.transaction_type === 'debit' ? 'Dr' : 'Cr',
        `Rs. ${tx.amount.toFixed(2)}`,
        `Rs. ${tx.running_balance.toFixed(2)}`
      ]);

      doc.autoTable({
        head: [tableColumn],
        body: tableRows,
        startY: 75,
        theme: 'striped',
        styles: { fontSize: 8, cellPadding: 3 },
        headStyles: { fillColor: [15, 23, 42], textColor: 255 },
        alternateRowStyles: { fillColor: [248, 250, 252] }
      });

      doc.save(`cashbook_statement_${format(new Date(), 'yyyyMMdd')}.pdf`);
      toast.success('PDF exported successfully');
    } catch (error) {
      toast.error('Failed to export PDF');
    } finally {
      setIsExportingPDF(false);
    }
  };

  return (
    <div className="flex gap-2">
      <Button variant="outline" className="bg-card" onClick={handleExportCSV} disabled={isExportingCSV || !transactions?.length}>
        {isExportingCSV ? <Loader2 className="w-4 h-4 mr-2 animate-spin" /> : <FileText className="w-4 h-4 mr-2" />}
        CSV
      </Button>
      <Button variant="outline" className="bg-card" onClick={handleExportPDF} disabled={isExportingPDF || !transactions?.length}>
        {isExportingPDF ? <Loader2 className="w-4 h-4 mr-2 animate-spin" /> : <Download className="w-4 h-4 mr-2" />}
        PDF
      </Button>
    </div>
  );
};

export default ExportButtons;
